File: week14_now_gimme_my_money/solutions/engine.c
// sum_engines.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <time.h>

#define N 5000000
#define MIN_CENTS 50
#define MAX_CENTS 200

void generate_values(double *values, int n) {
    srand(time(NULL));
    for (int i = 0; i < n; i++) {
        int cents = rand() % (MAX_CENTS - MIN_CENTS + 1) + MIN_CENTS;
        values[i] = cents / 100.0;
    }
}

double penny_engine(double *values, int n) {
    long total_pennies = 0;
    for (int i = 0; i < n; i++) {
        total_pennies += (long)(values[i] * 100);
    }
    return total_pennies / 100.0;
}

double float_engine(double *values, int n) {
    double total = 0.0;
    for (int i = 0; i < n; i++) {
        total += values[i];
    }
    return total;
}

double kahan_engine(double *values, int n) {
    double sum = 0.0;
    double c = 0.0;
    for (int i = 0; i < n; i++) {
        double y = values[i] - c;
        double t = sum + y;
        c = (t - sum) - y;
        sum = t;
    }
    return sum;
}

int main() {
    double *values = malloc(sizeof(double) * N);
    generate_values(values, N);

    double p_total = penny_engine(values, N);
    double f_total = float_engine(values, N);
    double k_total = kahan_engine(values, N);

    printf("Total using PennyEngine: %.10f\n", p_total);
    printf("Total using FloatMathEngine: %.10f\n", f_total);
    printf("Total using KahanEngine: %.10f\n", k_total);

    free(values);
    return 0;
}

// Compile with: gcc -o sum_engines sum_engines.c -lm
// Run with: ./sum_engines

File: week14_now_gimme_my_money/solutions/go/engine.js
function getValues(n) {
  const min = 50; // in cents
  const max = 200;
  const values = [];
  for (let i = 0; i < n; i++) {
    const cents = Math.floor(Math.random() * (max - min + 1)) + min;
    values.push(cents / 100.0);
  }
  return values;
}

// Penny-style accumulation
function pennyEngine(values) {
  let totalPennies = 0;
  for (let val of values) {
    totalPennies += Math.round(val * 100); // safer than implicit cast
  }
  return totalPennies / 100.0;
}

// Basic float sum
function floatMathEngine(values) {
  return values.reduce((acc, v) => acc + v, 0);
}

// Kahan Summation
function kahanEngine(values) {
  let sum = 0.0;
  let c = 0.0;
  for (let v of values) {
    let y = v - c;
    let t = sum + y;
    c = (t - sum) - y;
    sum = t;
  }
  return sum;
}

// Neumaier Summation
function neumaierEngine(values) {
  let sum = 0.0;
  let c = 0.0;
  for (let v of values) {
    let t = sum + v;
    if (Math.abs(sum) >= Math.abs(v)) {
      c += (sum - t) + v;
    } else {
      c += (v - t) + sum;
    }
    sum = t;
  }
  return sum + c;
}

// Run test
const n = 5_000_000;
const values = getValues(n);

console.log("Total using PennyEngine:       ", pennyEngine(values).toFixed(12));
console.log("Total using FloatMathEngine:  ", floatMathEngine(values).toFixed(12));
console.log("Total using KahanEngine:      ", kahanEngine(values).toFixed(12));
console.log("Total using NeumaierEngine:   ", neumaierEngine(values).toFixed(12));

//  to run:
//  node engine.js
//
